### PR TITLE
UPSTREAM: 30145: Add PVC storage to Limit Range

### PR DESCRIPTION
**Pull request by markturansky**
Cherry-pick of https://github.com/kubernetes/kubernetes/pull/30145

@abhgupta @derekwaynecarr @smarterclayton

### CODE REVIEW

**Comment by markturansky** [comment]
@pweil- @liggitt PTAL? Not sure who to ping for this cherry pick.

Docs added here:  https://github.com/openshift/openshift-docs/pull/3084

**Comment by pweil-** [comment]
@markturansky since this is going in to Kube 1.5 and 3.4 is based on 1.4 is there justification for merge? Ie, is this a critical feature for 3.4?

@smarterclayton thoughts on the above?  At this point I think we should be holding off cherry-picks as much as possible.

**Comment by markturansky** [comment]
Our Dedicated offering can get by with ClusterResourceQuota, which would quota storage by project. This solution works with the limited numbers in Dedicated.

In the Online version, CRQ won't scale as needed and so we can't use it. We need to limit the size of PVCs in a project in Online. We can already limit the count. This would limit overall consumption.

@abhgupta please confirm

**Comment by markturansky** [comment]
Also worth nothing, the Ops teams really want to get multi-zone volumes in Online.  3.3 supports multi-zone, but we can't because our custom provisioner is what provides our current storage consumption limits. So we're looking to kill our our custom provisioner in 3.4, but we need this cherry-pick.

**Comment by abhgupta** [comment]
@pweil- this is required for Online DevPreview and the upcoming paid tier that will be based on OCP 3.4. We have been hit with multiple issues maintaining the custom dynamic provisioner with Online and really need to be able to get rid of that and just leverage the product to provide the quota/limit restrictions. 

If this does not become part of OCP 3.4, then we will need to rebase our custom provisioner to ensure it works well with the stock PVC controllers (OCP 3.4), which poses additional challenges and delays at our end.

File: vendor/k8s.io/kubernetes/pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2544,6 +2544,17 @@ func ValidateLimitRange(limitRange *api.LimitRange) field.ErrorList {
 			}
 		}
 
+		if limit.Type == api.LimitTypePersistentVolumeClaim {
+			_, minQuantityFound := limit.Min[api.ResourceStorage]
```
Comment (derekwaynecarr) [inline vendor/k8s.io/kubernetes/pkg/api/validation/validation.go]: @markturansky - i must have missed this in the review upstream.  i should be able to set just a min and not a max, or a max and not a min, disagree?

Comment (markturansky) [inline vendor/k8s.io/kubernetes/pkg/api/validation/validation.go]: Allowing either to be empty opens the limit range to whatever the underlying infrastructure wants to enforce. IDK if that's good or bad, but AWS, for example, would enforce a 1Gi minimum and some large max Gi.

0 is an effective "no min".

Comment (derekwaynecarr) [inline vendor/k8s.io/kubernetes/pkg/api/validation/validation.go]: this is a confusing special case and I would prefer to get rid of it.

can we get rid of it here and make a separate pr to upstream that i can merge?

Comment (derekwaynecarr) [inline vendor/k8s.io/kubernetes/pkg/api/validation/validation.go]: we can loosen the validation upstream and merge this now.

**Comment by derekwaynecarr** [comment]
[merge]

**Comment by abhgupta** [comment]
re-[merge] Failure in `TestRegistryClientAPIv1` seems like a fluke

**Comment by derekwaynecarr** [comment]
FYI - 3.4 has the ability quota cumulative requests.storage and not just
number of claims, but I assume you want this to control the size of an
individual claim.  If it lets us get rid of other tech debt, I am fine
merging this now.  I agree that we should challenge all things like this in
the future though.

**Pull request merged**